### PR TITLE
#159965306 implement CRUD for articles

### PR DESCRIPTION
**Pull request by meshack-mbuvi**
#### What does this PR do?
- Have CRUD on articles endpoint in Authors Haven REST API.
#### Description of Task to be completed?
- Ensure the following endpoints are working:
    - GET /api/articles/
    - GET /api/articles/{id}
    - POST /api/articles/
    - PUT /api/articles/{id}
    - DELETE /api/articles/{id}
#### How should this be manually tested?
- Clone [this](https://github.com/andela/ah-magnificent6.git) repository
- Create and activate a [virtual environment](https://docs.python.org/3/library/venv.html) inside the project base directory.
- Run the command `git checkout ft-crud-on-articles-159965306`
- Install the project dependencies by running `pip install -r requirements.txt`
- Run the server by issuing the commant `python manage.py runserver`
- Using ([Postman](https://www.getpostman.com/) or your favorite REST API client, 
- Create an account by accessing `api/users/signup/` by sending the payload in the format 
`
{
  "firstname": "admin",
  "username": "admin",
  "email": "admin@gmail.com",
  "password": "password",
} 
`
and 
login on `/api/users/login/` using the same format for the payload above
- test the above endpoints by passing data in the format 
`{
    'title':'title of the article',
    'body':'body of the article',
    'description':'Description of the article'
}`

#### Any background context you want to provide?
 In Authors Haven platform, users will have the capability to create articles and publish them for other people to read. By having this ability, users can showcase their creativity and thoughts by creating and publishing articles in our platform.
#### What are the relevant pivotal tracker stories?
[#159965306 User should be able to create articles](https://www.pivotaltracker.com/story/show/159965306)
#### Screenshots 
post an article
image
Retrieve all articles
image

Retrieve a single article
image

edit an article
image

Delete an article
image

### CODE REVIEW

File: authors/apps/articles/models.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    favorited = models.NullBooleanField(default=False)
+    favoritesCount = models.IntegerField(default=0)
+    image = models.ImageField(
+        upload_to='static/images', default='static/images/no-img.jpg')
+
+    def __str__(self):
+        "Returns a string representation of article title."
+        return self.title
+
+    def _generate_unique_slug(self):
```
Comment (parseen254) [inline authors/apps/articles/models.py]: @meshack-mbuvi You might be reinventing the wheel here. https://books.agiliq.com/projects/django-orm-cookbook/en/latest/slugfield.html check this article out.

File: .travis.yml [hunk 2]
```diff
@@ -8,19 +8,18 @@ install:
...
   - coverage erase
-  - coverage run --source='.' manage.py test
+  - coverage run --source=authors/apps manage.py test
   - coverage report -m
   - coveralls
 after_success:
   - coveralls
 
-deploy:
+? deploy
```
Comment (parseen254) [inline .travis.yml]: @meshack-mbuvi what's happening here?

File: authors/apps/articles/renderers.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import json
+
+from rest_framework.renderers import JSONRenderer
+from rest_framework.utils.serializer_helpers import ReturnDict
+
+
+class ArticleJSONRenderer(JSONRenderer):
+    charset = 'utf-8'
```
Comment (parseen254) [inline authors/apps/articles/renderers.py:8]: Also here checkout http://www.django-rest-framework.org/api-guide/renderers/

File: authors/apps/articles/views.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+class ArticleAPIView(generics.ListCreateAPIView):
+    """
+    user generics.CreateAPIView to expose parameters to the documentation
+    method: post
+    Create a new article
+    """
+    queryset = Article.objects.all()
+    serializer_class = ArticleSerializer
+    permission_classes = (IsAuthenticated,)
+    renderer_classes = (ArticleJSONRenderer,)
```
Comment (parseen254) [inline authors/apps/articles/views.py]: Check out inbuilt JSON renderer.

File: authors/apps/articles/views.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        }
+        # pass article data to the serializer class, check whether the data is
+        # valid and if valid, save it.
+        serializer = self.serializer_class(data=article)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+        return Response(serializer.data, status.HTTP_201_CREATED)
+
+
+class ArticleRetrieveUpdateDestroy(generics.RetrieveUpdateDestroyAPIView):
```
Comment (parseen254) [inline authors/apps/articles/views.py]: Rename this to article details view.

File: requirements.txt [hunk 6]
```diff
@@ -1,10 +1,29 @@
+autopep8==1.4
```
Comment (parseen254) [inline requirements.txt]: let us keep our requirements clean.

**Review by parseen254 — CHANGES_REQUESTED** [review]

**Comment by meshack-mbuvi** [comment]
On it.

File: .travis.yml (hunk 2 above)
Comment (meshack-mbuvi) [reply .travis.yml]: Whenever I save the file, the editor is turning that statement to look as you have seen above.But am going to have a workaround about it.

File: requirements.txt (hunk 6 above)
Comment (meshack-mbuvi) [reply requirements.txt]: Noted.

**Review by parseen254 — APPROVED** [review]

**Pull request merged**